python/shotgun_data/shotgun_data_retriever.py

```diff
@@ -0,0 +1,716 @@
+                                              priority, 
+                                              group = self._bg_tasks_group,
+                                              task_args = task_args,
+                                              task_kwargs = task_kwargs)
+        return str(task_id)
+
+
+    def request_thumbnail(self, url, entity_type, entity_id, field, load_image=False):
+        """
+        Downloads a thumbnail form Shotgun asyncronously or returns the cached thumbnail 
+        if found.  Optionally loads the thumbnail into a QImage.
+
+        :param url:         The thumbnail url string that is associated with this thumbnail. This is
+                            the field value as returned by a Shotgun query.
+        :param entity_type: Shotgun entity type with which the thumb is associated.
+        :param entity_id:   Shotgun entity id with which the thumb is associated.
+        :param field:       Thumbnail field. Normally 'image' but could also for example be a deep 
+                            link field such as 'sg_sequence.Sequence.image'
+        :param load_image:  If set to True, the return data structure will contain a QImage object 
+                            with the image data loaded.
+
+        :returns:           A unique task id representing this request.
+        """
+        if not self._task_manager:
+            return
+
+        # always add check for thumbnail already downloaded:
+        check_task_id = self._task_manager.add_task(self._task_check_thumbnail,
+                                                    priority = ShotgunDataRetriever._CHECK_THUMB_PRIORITY,
+                                                    group = self._bg_tasks_group,
+                                                    task_kwargs = {"url":url, 
+                                                                   "load_image":load_image})
+
+        # add download thumbnail task
+        dl_task_id = self._task_manager.add_task(self._task_download_thumbnail,
+                                                 upstream_task_ids = [check_task_id],
+                                                 priority = ShotgunDataRetriever._DOWNLOAD_THUMB_PRIORITY,
+                                                 group = self._bg_tasks_group,
+                                                 task_kwargs = {"url":url,
+                                                                "entity_type":entity_type, 
+                                                                "entity_id":entity_id,
+                                                                "field":field,
+                                                                "load_image":load_image})
+
+        # all results for requesting a thumbnail should be returned with the same id so use
+        # a mapping to track the 'primary' task id:
+        self._thumb_task_id_map[dl_task_id] = check_task_id
+
+        return str(check_task_id)
```

Comment:
are we casting the task ids to strings to be more backwards compatible? (old version of api returned guid strings) I am worried that it might be easy to miss doing this cast - wondering if we should just return the ints without casting? It's a major version bump so it's ok with things breaking, so maybe better to be consistent between the return values here and the behaviour of the new task manager API - so that they work the same way?

Also, quick question - previously the typical pattern was to stash these task guids as keys in a dict to then process things later - will everything be guaranteed to be unique with the new setup as well - even with ints? Could there be a case where you somehow could end up with two task ids being the same somehow? I am guessing no but thought it's worth just asking!